// @tanstack/devtools-vite/dist/esm/utils.js
import fs from "node:fs/promises";
import { normalizePath } from "vite";
const handleDevToolsViteRequest = (req, res, next, cbOrOptions) => {
  const options = typeof cbOrOptions === "function" ? { onOpenSource: cbOrOptions } : cbOrOptions;
  if (req.url?.includes("__tsd/open-source")) {
    const searchParams = new URLSearchParams(req.url.split("?")[1]);
    const source = searchParams.get("source");
    if (!source) {
      return;
    }
    const parsed = parseOpenSourceParam(source);
    if (!parsed) {
      return;
    }
    const { file, line, column } = parsed;
    options.onOpenSource?.({
      type: "open-source",
      routine: "open-source",
      data: {
        source: file ? normalizePath(`${process.cwd()}/${file}`) : void 0,
        line,
        column
      }
    });
    res.setHeader("Content-Type", "text/html");
    res.write(`<script> window.close(); <\/script>`);
    res.end();
    return;
  }
  if (req.url?.includes("__tsd/console-pipe/sse") && req.method === "GET") {
    if (options.onConsolePipeSSE) {
      options.onConsolePipeSSE(res, req);
      return;
    }
    return next();
  }
  if (req.url?.includes("__tsd/console-pipe/server") && req.method === "POST") {
    if (options.onServerConsolePipe) {
      let body = "";
      req.on("data", (chunk) => {
        body += chunk.toString();
      });
      req.on("end", () => {
        try {
          const { entries } = JSON.parse(body);
          options.onServerConsolePipe(entries);
          res.statusCode = 200;
          res.end("OK");
        } catch {
          res.statusCode = 400;
          res.end("Bad Request");
        }
      });
      return;
    }
    return next();
  }
  if (req.url?.includes("__tsd/console-pipe") && req.method === "POST") {
    if (options.onConsolePipe) {
      let body = "";
      req.on("data", (chunk) => {
        body += chunk.toString();
      });
      req.on("end", () => {
        try {
          const { entries } = JSON.parse(body);
          options.onConsolePipe(entries);
          res.statusCode = 200;
          res.end("OK");
        } catch {
          res.statusCode = 400;
          res.end("Bad Request");
        }
      });
      return;
    }
    return next();
  }
  if (!req.url?.includes("__tsd")) {
    return next();
  }
  const chunks = [];
  req.on("data", (chunk) => {
    chunks.push(chunk);
  });
  req.on("end", () => {
    const dataToParse = Buffer.concat(chunks);
    try {
      const parsedData = JSON.parse(dataToParse.toString());
      options.onOpenSource?.(parsedData);
    } catch (e) {
    }
    res.write("OK");
    res.end();
  });
};
const parseOpenSourceParam = (source) => {
  const parts = source.match(/^(.+):(\d+):(\d+)$/);
  if (!parts) return null;
  const [, file, line, column] = parts;
  return { file, line, column };
};
const tryReadFile = async (filePath) => {
  try {
    const data = await fs.readFile(filePath, "utf-8");
    return data;
  } catch (error) {
    return null;
  }
};
const tryParseJson = (jsonString) => {
  if (!jsonString) {
    return null;
  }
  try {
    const result = JSON.parse(jsonString);
    return result;
  } catch (error) {
    return null;
  }
};
const readPackageJson = async () => tryParseJson(await tryReadFile(process.cwd() + "/package.json"));
const stripEnhancedLogPrefix = (args, formatSourceLocation) => {
  if (args.length === 0) return args;
  let sourceArgIndex = -1;
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (typeof arg === "string" && arg.includes("__tsd/open-source?source=")) {
      sourceArgIndex = i;
      break;
    }
  }
  if (sourceArgIndex === -1) {
    return args;
  }
  const sourceArg = args[sourceArgIndex];
  let sourceLocation = "";
  const sourceMatch = sourceArg.match(/source=([^&\s]+?)%c/);
  if (sourceMatch?.[1]) {
    try {
      sourceLocation = decodeURIComponent(sourceMatch[1]);
      if (sourceLocation.startsWith("/")) {
        sourceLocation = sourceLocation.slice(1);
      }
    } catch {
    }
  }
  const styleCount = (sourceArg.match(/%c/g) || []).length;
  const userArgsStart = sourceArgIndex + 1 + styleCount;
  const result = [];
  if (sourceLocation) {
    result.push(
      formatSourceLocation ? formatSourceLocation(sourceLocation) : sourceLocation
    );
  }
  for (let i = userArgsStart; i < args.length; i++) {
    result.push(args[i]);
  }
  return result.length > 0 ? result : args;
};
export {
  handleDevToolsViteRequest,
  parseOpenSourceParam,
  readPackageJson,
  stripEnhancedLogPrefix,
  tryParseJson
};
